src/FAudio_platform_win32.c

```diff
@@ -202,6 +203,128 @@ static DWORD WINAPI FAudio_AudioClientThread(void *user)
 	return 0;
 }
 
+/* Returns the device index for the default device.
+ * Returns 0 on failure. This will prevent the rest of the FAudio code from 
+ * re-ordering the devices. */
+static uint32_t FAudio_DefaultDeviceIndex(IMMDeviceCollection *device_collection)
+{
+	IMMDevice *device;
+	HRESULT hr;
+	uint32_t i, count;
+	WCHAR *default_guid;
+	WCHAR *device_guid; 
+
+	/* Open the default device and get its GUID. */
+	hr = IMMDeviceEnumerator_GetDefaultAudioEndpoint(
+		device_enumerator,
+		eRender,
+		eConsole,
+		&device
+	);
+	if (FAILED(hr))
+	{
+		return 0;
+	}
+	hr = IMMDevice_GetId(device, &default_guid);
+	if (FAILED(hr))
+	{
+		IMMDevice_Release(device);
+		return 0;
+	}
+
+	/* Free the default device. */
+	IMMDevice_Release(device);
+
+	hr = IMMDeviceCollection_GetCount(device_collection, &count);
+	if (FAILED(hr))
+	{
+		CoTaskMemFree(default_guid);
+		return 0;
+	}
+
+	for (i = 0; i < count; i += 1)
+	{
+		/* Open the device and get its GUID. */
+		hr = IMMDeviceCollection_Item(device_collection, i, &device);
+		if (FAILED(hr)) {
+			CoTaskMemFree(default_guid);
+			return 0;
+		}
+		hr = IMMDevice_GetId(device, &device_guid);
+		if (FAILED(hr))
+		{
+			CoTaskMemFree(default_guid);
+			IMMDevice_Release(device);
+			return 0;
+		}
+
+		if (lstrcmpW(default_guid, device_guid) == 0)
+		{
+			/* Device found. */
+			CoTaskMemFree(default_guid);
+			CoTaskMemFree(device_guid);
+			IMMDevice_Release(device);
+			return i;
+		}
+
+		CoTaskMemFree(device_guid);
+		IMMDevice_Release(device);
+	}
+
+	/* This should probably never happen. */
+	CoTaskMemFree(default_guid);
+	return 0;
+}
+
+/* Open `device`, corresponding to `deviceIndex`. `deviceIndex` 0 always
+ * corresponds to the default device. XAudio reorders the devices so that the
+ * default device is always at index 0, so we mimick this behavior here by
+ * swapping the devices at indexes 0 and `defaultDeviceIndex`.
+ */
+static DWORD WINAPI FAudio_OpenDevice(uint32_t deviceIndex, IMMDevice **device)
```

Comment:
Should probably return `HRESULT`, and `WINAPI` isn't needed so it should be dropped.

Reply:
Done.